pkg/perf/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,97 @@
+package healthcheck
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg/app"
+	"github.com/threefoldtech/zos/pkg/perf"
+	"github.com/threefoldtech/zos/pkg/stubs"
+)
+
+const (
+	id          = "healthcheck"
+	schedule    = "0 */20 * * * *"
+	description = "health check task runs multiple checks to ensure the node is in a usable state and set flags for the power daemon to stop reporting uptime if it is not usable"
+)
+
+// NewTask returns a new health check task.
+func NewTask() perf.Task {
+	checks := []checkFunc{
+		cacheCheck,
+	}
+	return &healthcheckTask{
+		checks: checks,
+		errors: make(map[string][]string),
+	}
+}
+
+type checkFunc func(context.Context) (string, error)
+
+type healthcheckTask struct {
+	checks []checkFunc
+	errors map[string][]string
+}
+
+var _ perf.Task = (*healthcheckTask)(nil)
+
+// ID returns task ID.
+func (h *healthcheckTask) ID() string {
+	return id
+}
+
+// Cron returns task cron schedule.
+func (h *healthcheckTask) Cron() string {
+	return schedule
+}
+
+// Description returns task description.
+func (h *healthcheckTask) Description() string {
+	return description
+}
+
+// Run executes the health checks.
+func (h *healthcheckTask) Run(ctx context.Context) (interface{}, error) {
+	log.Debug().Msg("starting health check task")
+	for k := range h.errors {
+		// reset errors on each run
+		h.errors[k] = make([]string, 0)
+	}
+
+	for _, check := range h.checks {
+		label, err := check(ctx)
+		if err == nil {
+			continue
+		}
+		h.errors[label] = append(h.errors[label], err.Error())
+	}
+
+	cl := perf.GetZbusClient(ctx)
+	zui := stubs.NewZUIStub(cl)
+
+	for label, data := range h.errors {
+		zui.PushErrors(ctx, label, data)
+	}
+	return h.errors, nil
+}
+
+func cacheCheck(ctx context.Context) (string, error) {
```

Comment:
returning the label is weird imho. The better approach is that the `checks` on the task itself can instead be a map of the (label, checkFunc)
```go
checks := map[string]checkFunc{
   "cache", cacheCheck
}
```
which means, your funcion can just return an error, and the loop will know what label to use

Reply:
I get what you mean, and the idea itself is good. But i still find it weird that the function has to return the label.

Other possible way to get the same effect is 
- `map[string][]checkFunc`  
- `checkFunc() []error` where the checkFunc can return a list of errors which means it internally can run mutliple checks and return a list of possible health errors.

```
health() []error {
   var issues []error
   if err := cacheCheck() err = !nil {
        issues = append(issues, err)
   }
   // more checks

  return issues
}
```